### DISCUSSION TITLE
How to Inject non registered monobehavior into plain c# class presenter?

### DISCUSSION THREAD

**Discussion by artahc** [hadashiA/VContainer#281] [Q&A]
Hi, I'm super new to DI in general, especially in Unity. 
I'm a little bit confused how things works
Currently I'm trying to implement MVP pattern in unity

So i have this monobehaviour class called MovementComponent , which implementing IMovable interface. 
```
public class MovementComponent : MonoBehaviour, IMovable {
	public void Move (Vector3 position, Quaternion rotation) {
		transform.SetPositionAndRotation (transform.position + position, Quaternion.Lerp (transform.rotation, rotation, Time.deltaTime * 10));
	}
}
```

And this is my presenter for MovementComponent:
```
public class MovementPresenter : IStartable {

	private readonly IMovable component;
	private readonly InputSource inputSource;

	public MovementPresenter (IMovable component, InputSource inputSource) {
		this.component = component;
		this.inputSource = inputSource;
		Debug.Log ($"{this} constructor.");
	}

	public void Start () {
		Observable.EveryUpdate ()
			.Select (_ => inputSource.OnMove.Value).Subscribe (input => {
				var moveDirection = new Vector3 (input.x, 0, input.y) * 10 * Time.deltaTime;
				var rotation = moveDirection != Vector3.zero ? Quaternion.LookRotation (moveDirection.normalized) : Quaternion.identity;
				component.Move (moveDirection, rotation);
			});
	}
}
```

My question is: how am I supposed to inject the IMoveable component into the presenter without registering the component directly on public/serializefield?

**Comment by hadashiA** [comment]
Short answer: There is no way to do that.

Depending on why you do not want to register MonoBehaviour, the solution may be different.

- If you have many instances of IMovable:
    -  The basic concept of DI is to identify types as keys. The reason for this is that it is recommended to associate services (functions) with each other rather than with value objects (Data oriented objects).
    - The solution is to give a special type to the object you want to control with inputSource, or to register a collection type of `IMovable`.

**Reply by artahc** [reply]
Hi thank you for replying. 
Actually what i was trying to achieve is I want to instantiate lets say character dynamically at runtime, multiple times (there will be few character in the scene), therefore i cant register the component in the inspector. 

However, i come up with the solution by creating IInstaller for the character, so when i need to instantiate a character, it would register the specified component for each presenter. But this will result having multiple scope with same registration, is this how its supposed works or am i wrong / misunderstand something?
